examples/chip-tool/main.cpp

```diff
@@ -16,36 +16,32 @@
  *
  */
 
-#include "commands/Commands.h"
-
-using namespace ::chip;
-using namespace ::chip::DeviceController;
+#include "commands/Echo.h"
+#include "commands/EchoBle.h"
+#include "commands/clusters/Identify/Commands.h"
+#include "commands/clusters/OnOff/Commands.h"
 
 // NOTE: Remote device ID is in sync with the echo server device id
 //       At some point, we may want to add an option to connect to a device without
 //       knowing its id, because the ID can be learned on the first response that is received.
-constexpr NodeId kLocalDeviceId  = 112233;
-constexpr NodeId kRemoteDeviceId = 12344321;
+constexpr chip::NodeId kLocalDeviceId  = 112233;
+constexpr chip::NodeId kRemoteDeviceId = 12344321;
+
+int run(int argc, char * argv[]);
 
 // ================================================================================
 // Main Code
 // ================================================================================
 int main(int argc, char * argv[])
 {
-    CHIP_ERROR err = CHIP_NO_ERROR;
-    ChipDeviceController dc;
-
-    err = dc.Init(kLocalDeviceId);
-    VerifyOrExit(err == CHIP_NO_ERROR, ChipLogError(Controller, "Init failure: %s", ErrorStr(err)));
-
-    err = dc.ServiceEvents();
-    VerifyOrExit(err == CHIP_NO_ERROR, ChipLogError(Controller, "Init Run Loop failure: %s", ErrorStr(err)));
-
-    err = RunCommand(&dc, kRemoteDeviceId, argc, argv);
-    SuccessOrExit(err);
-
-exit:
-    dc.ServiceEventSignal();
-    dc.Shutdown();
-    return (err == CHIP_NO_ERROR) ? EXIT_SUCCESS : EXIT_FAILURE;
+    Echo echo;
+    EchoBle echoBle;
+    Off off;
+    On on;
+    ReadOnOff readOnOff;
+    Toggle toggle;
+    Identify identify;
+    IdentifyQuery identifyQuery;
```

Comment:
There's still a list here and now we have a global variable, too. Couldn't we do

```
Commands commands;
commands.Register(std::make_unique<Echo>());
// etc
```

to avoid the globals?

Or maybe

```
using CreateCommandFn = std::unique_ptr<Command>();
std::vector<CreateCommandFn> commandsList = {
  &Echo::Create,
  &EchoBle::Create,
  &Off::Create,
  &On::Create,
  &ReadOnOff::Create,
  &Toggle::Create,
  &Identify::Create,
  &IdentifyQuery::Create
};

Commands commands(commandsList);
```

and add static `Command::Create` functions to all the commands.

I don't see why we need a global vector of commands that points into the stack.

Reply:
I rewrite the code to use something similar to:
```
Commands commands;
commands.Register(std::make_unique<Echo>());
```

But instead of having a single place to register cluster commands, I have splitter it into helpers in the `clusters/*/Commands.h` files.

So in `main.cpp`the code looks like:
```
Commands commands;
registerClusterOnOff(commands);
registerClusterIdentify(commands);
```

And in `clusters/OnOff/Commands.h`, the helper method looks like:
```
void registerClusterOnOff(Commands & commands)
{
    commands.Register(make_unique<Off>(cluster_id));
    commands.Register(make_unique<On>(cluster_id));
    commands.Register(make_unique<ReadOnOff>(cluster_id));
    commands.Register(make_unique<Toggle>(cluster_id));
}
```